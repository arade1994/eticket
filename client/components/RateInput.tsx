import { Dispatch, SetStateAction } from "react";

interface Props {
  comment: string;
  setComment: Dispatch<SetStateAction<string>>;
  setRate: Dispatch<SetStateAction<number>>;
}

const RateInput = ({ comment, setComment, setRate }: Props) => {
  return (
    <>
      <div className="rating">
        {Array.from(Array(5).keys())
          .sort((a, b) => b - a)
          .map((index) => (
            <>
              <input
                type="radio"
                id={`star${index}`}
                name="rating"
                value={index}
              />
              <label
                htmlFor={`star${index}`}
                className="star"
                onClick={() => setRate(index + 1)}
              >
                &#9733;
              </label>
            </>
          ))}
      </div>
      <div className="comment">
        <label htmlFor="comment">Tell us more:</label>
        <br />
        <textarea
          id="comment"
          name="comment"
          value={comment}
          onChange={(e) => setComment(e.target.value)}
        ></textarea>
      </div>
    </>
  );
};

export default RateInput;
