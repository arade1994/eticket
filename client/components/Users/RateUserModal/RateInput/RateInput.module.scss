.rateInputContainer {
  width: "100%";
  display: flex;
  flex-direction: column;
  justify-content: space-evenly;
  text-align: center;

  & .rate {
    margin-bottom: 20px;
    display: flex;
    flex-direction: row-reverse;
    justify-content: center;

    & input {
      display: none;
    }

    & .star {
      font-size: 24px;
      cursor: pointer;
    }

    & .star:hover,
    & .star:hover ~ .star {
      color: orange;
    }

    & input:checked ~ .star {
      color: orange;
    }

    & .checked {
      color: orange;
    }
  }

  & .comment {
    margin-bottom: 20px;

    & label {
      opacity: 0.5;
    }

    & textarea {
      width: 100%;
      height: 100px;
      resize: none;
      border: none;
    }
  }
}
