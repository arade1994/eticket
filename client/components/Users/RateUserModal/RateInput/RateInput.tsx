import { ChangeEvent } from "react";
import classes from "./RateInput.module.scss";

interface Props {
  comment: string;
  onChangeComment: (event: ChangeEvent<HTMLTextAreaElement>) => void;
  onChangeRate: (value: number) => void;
}

const RateInput = ({ comment, onChangeComment, onChangeRate }: Props) => {
  return (
    <div className={classes.rateInputContainer}>
      <div className={classes.rate}>
        {Array.from(Array(5).keys())
          .sort((a, b) => b - a)
          .map((index) => (
            <>
              <input
                type="radio"
                id={`star${index}`}
                name="rating"
                value={index}
              />
              <label
                htmlFor={`star${index}`}
                className={classes.star}
                onClick={() => onChangeRate(index + 1)}
              >
                &#9733;
              </label>
            </>
          ))}
      </div>
      <div className={classes.comment}>
        <label htmlFor="comment">Tell us more:</label>
        <br />
        <textarea
          id="comment"
          name="comment"
          value={comment}
          onChange={onChangeComment}
        ></textarea>
      </div>
    </div>
  );
};

export default RateInput;
