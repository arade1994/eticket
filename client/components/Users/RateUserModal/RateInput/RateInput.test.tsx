import { cleanup, render, screen } from "@testing-library/react";
import { afterEach, describe, expect, test, vi } from "vitest";
import { userEvent } from "@testing-library/user-event";
import RateInput from "./RateInput";

const setCommentMock = vi.fn();
const setRateMock = vi.fn();

afterEach(() => cleanup());

describe("<RateInput />", () => {
  test("It should render rate input component as expected", async () => {
    const { baseElement } = render(
      <RateInput
        comment="Test comment"
        onChangeComment={setCommentMock}
        onChangeRate={setRateMock}
      />
    );

    expect(baseElement).toMatchSnapshot();

    const comment = await screen.findByRole("textbox");
    expect(comment).toBeDefined();
    expect(comment.textContent).toEqual("Test comment");
  });

  test("It should call setComment() method when comment is changed", async () => {
    render(
      <RateInput
        comment="Test comment"
        onChangeComment={setCommentMock}
        onChangeRate={setRateMock}
      />
    );

    const commentTxtArea = (await screen.findByRole(
      "textbox"
    )) as HTMLTextAreaElement;
    userEvent.clear(commentTxtArea);
    await userEvent.type(commentTxtArea, "Updated comment");
    expect(setCommentMock).toHaveBeenCalled();
  });

  test("It should call setRate() method when rate is selected", async () => {
    const { container } = render(
      <RateInput
        comment="Test comment"
        onChangeComment={setCommentMock}
        onChangeRate={setRateMock}
      />
    );

    const star3 = container.querySelector("[for='star3']") as Element;
    await userEvent.click(star3);
    expect(setRateMock).toHaveBeenCalled();
  });
});
