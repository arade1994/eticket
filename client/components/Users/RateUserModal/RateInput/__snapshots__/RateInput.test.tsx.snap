// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RateInput /> > It should render rate input component as expected 1`] = `
<body>
  <div>
    <div
      class="_rateInputContainer_c17dbd"
    >
      <div
        class="_rate_c17dbd"
      >
        <input
          id="star4"
          name="rating"
          type="radio"
          value="4"
        />
        <label
          class="_star_c17dbd"
          for="star4"
          id="label-star-4"
        >
          ★
        </label>
        <input
          id="star3"
          name="rating"
          type="radio"
          value="3"
        />
        <label
          class="_star_c17dbd"
          for="star3"
          id="label-star-3"
        >
          ★
        </label>
        <input
          id="star2"
          name="rating"
          type="radio"
          value="2"
        />
        <label
          class="_star_c17dbd"
          for="star2"
          id="label-star-2"
        >
          ★
        </label>
        <input
          id="star1"
          name="rating"
          type="radio"
          value="1"
        />
        <label
          class="_star_c17dbd"
          for="star1"
          id="label-star-1"
        >
          ★
        </label>
        <input
          id="star0"
          name="rating"
          type="radio"
          value="0"
        />
        <label
          class="_star_c17dbd"
          for="star0"
          id="label-star-0"
        >
          ★
        </label>
      </div>
      <div
        class="_comment_c17dbd"
      >
        <label
          for="comment"
        >
          Tell us more:
        </label>
        <br />
        <textarea
          id="comment"
          name="comment"
        >
          Test comment
        </textarea>
      </div>
    </div>
  </div>
</body>
`;
